design/nfs/nfs-quota.md

```diff
@@ -0,0 +1,120 @@
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: rook-nfs-pv-claim
+spec:
+  storageClassName: "rook-nfs-share"
+  accessModes:
+    - ReadWriteMany
+  resources:
+    requests:
+      storage: 1Mi
+```
+
+The client still can use the higher capacity than `1mi` as requested.
+
+This proposal is to add features which the Rook NFS Provisioner will provide the specific capacity as requested from `.spec.resources.requests.storage` field in PersistentVolumeClaim.
+
+## Implementation
+
+The implementation will be use `Project Quota` on xfs filesystem. When the users need to use the quota feature they should use xfs filesystem with `prjquota/pquota` mount options for underlying volume. Users can specify filesystem type and mount options through StorageClass that will be used for underlying volume. For example:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard-xfs
+parameters:
+  fsType: xfs
+mountOptions:
+  - prjquota
+...
+```
+
+Then the underlying PersistentVolumeClaim should be used that StorageClass 
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: nfs-default-claim
+spec:
+  storageClassName: "standard-xfs"
+  accessModes:
+  - ReadWriteOnce
+...
+```
+
+If the above conditions are met then the Rook NFS Provisioner will create projects and set the quota limit using [xfs_quota](https://linux.die.net/man/8/xfs_quota) before creating PersistentVolumes based on `.spec.resources.requests.storage` field in PersistentVolumeClaim. Otherwise the Rook NFS Provisioner will provision a PersistentVolumes without creating setting the quota.
+
+To creating the project, Rook NFS Provisioner will invoke the following command
+
+> xfs_quota -x -c project -s -p '*nfs_pv_directory* *project_id*' *projects_file*
+
+And setting quota with the command
+
+> xfs_quota -x -c 'limit -p bhard=*size* *project_id*' *projects_file*
+
+which
+
+1. *nfs_pv_directory* is sub-directory from exported directory that used for NFS PV.
+1. *project_id* is unique id `uint16` 1 to 65535.
+1. *size* is size of quota as requested.
+1. *projects_file* is file that contains *project quota block* for persisting quota state purpose. In case the Rook NFS Provisioner pod is killed, Rook NFS Provisioner pod will restore the quota state based on *project quota block* entries in *projects_file* at startup.
+1. *project quota block* is combine of *project_id*:*nfs_pv_directory*:*size*
+
+Since Rook NFS has possibility to create more than one of nfs share/export that have different underlying volume directory, the *projects_file* will saved on each underlying volume directory. So each nfs share/export will have different *projects_file* and *project_file* will still persist. The *projects_file* will ony created if underlying volume directory mounted as xfs with prjquota mount options. This mean the existence of *project_file* will indicate if quota was enabled. The hierarchy of directory will look like
```

Comment:
```suggestion
Since Rook NFS has the ability to create more than one NFS share/export that have different underlying volume directories, the *projects_file* will be saved on each underlying volume directory. So each NFS share/export will have different *projects_file* and each *project_file* will be persisted. The *projects_file* will only be created if underlying volume directory is mounted as `xfs` with `prjquota` mount options. This mean the existence of *project_file* will indicate if quota was enabled. The hierarchy of directory will look like:
```